torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -402,6 +429,58 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
   return futureResponseMessage;
 }
 
+void TensorPipeAgent::pollTimeoutRpcs() {
+  while (rpcAgentRunning_.load()) {
+    std::unique_lock<std::mutex> lock(timeoutMapMutex_);
+
+    steady_clock_time_point earliestTimeout =
+        std::chrono::steady_clock::now() + kLargeTimeDuration;
+
+    // We sleep until the earliest expiring RPC in the timeoutMap_. We must
+    // also ensure that we sleep while the map is empty, and we exit sleeping
+    // if the RPC Agent has been shutdown.
+    for (;;) {
+      if (!rpcAgentRunning_.load()) {
+        return;
+      }
+      if (std::chrono::steady_clock::now() >= earliestTimeout) {
+        break;
+      }
+      if (!timeoutMap_.empty()) {
+        earliestTimeout = timeoutMap_.begin()->first;
+      }
+      timeoutThreadCV_.wait_until(lock, earliestTimeout);
+    }
+
+    // Move all these futures to a separate vector so we can process them
+    // outside the lock.
+    std::vector<std::shared_ptr<FutureMessage>> timedOutFutures =
+        std::move(timeoutMap_.begin()->second);
+    // We can safely remove this key from the timeoutMap_ since all these
+    // futures will be processed.
+    timeoutMap_.erase(timeoutMap_.begin());
+
+    lock.unlock();
+
+    // Set an error on futures added to the timedOutFutures vector. We do this
+    // outside the lock to prevent potential lock-order-inversions by callbacks
+    // triggered by the serError call.
+    for (const auto& future : timedOutFutures) {
+      std::string errorMsg = c10::str(
+          "RPC ran for more than set timeout and will now be marked with an error");
+      // Using setErrorIfNeeded so completed futures are ignored.
+      future->setErrorIfNeeded(errorMsg);
+      if (future->hasError()) {
+        // We only want to decrease the active call count if the future has an
+        // error. If it was successful, the counter was already updated in the
+        // send() function above.
+        --clientActiveCalls_;
+      }
```

Comment:
I still think this could lead to wrong counts: if the future had already been marked with an error (say, because of an I/O problem), then the count had already been decreased, and yet here we would decrease it a second time.

I see several ways to fix this:
- change the `setErrorIfNeeded` method of the torch::utils::Future to return a bool indicating whether the error was actually set (or whether the future was already complete)
- As soon as we create the Future in the send method, we call `addCallback` on it with a lambda that decreases `clientActiveCalls_`. The future ensures that such callbacks are only called once, the first time the future is marked complete (whether successfully or with an error), and therefore guarantees no double-decrease
- We could change the other code paths where we complete a future (also both for success and error outcomes) so that they remove the future from the `timeoutMap_`

I think the second solution is by far the simpler and more robust one.

Reply:
Implemented option 2 in the next PR in this stack (to modularize the changes a bit).

Reply:
Thanks, looks good! And, as always, it's good to split up changes as much as it's reasonable. One tip though: when exposing a bug while writing a diff, I'd tend to fix it by _prepending_ a diff to the stack, rather than appending it. This way each commit is bug-free, and this is a good invariant to preserve, as it allows for example a bisect to run smoothly. In this case the bug introduced by this diff is minor (just a count that may sometimes be off), so it's fine.